tools/spawn/spawn.c

```diff
@@ -0,0 +1,414 @@
+
+	int num_of_forks;
+	struct fork_info *fork_list;
+
+	pthread_mutex_t mt;
+};
+
+static struct spawn_opt g_opt;
+
+/*
+ * show usage
+ */
+
+static void usage(void)
+{
+	fprintf(stderr, "Usage: %s <timeout in seconds> <number of forks> <command file>\n",
+						SPAWN_PROG_NAME);
+	exit(-1);
+}
+
+/*
+ * init options
+ */
+
+static void init_options(void)
+{
+	memset(&g_opt, 0, sizeof(g_opt));
+	pthread_mutex_init(&g_opt.mt, NULL);
+}
+
+/*
+ * finalize options
+ */
+
+static void finalize_options(void)
+{
+	int i;
+
+	/* wait until forking threads has been terminated */
+	for (i = 0; i < g_opt.num_of_forks; i++) {
+		if (g_opt.fork_list[i].th < 0)
+			continue;
+
+		pthread_join(g_opt.fork_list[i].th, NULL);
+	}
+
+	/* free forking info list */
+	free(g_opt.fork_list);
+
+	/* destroy mutex */
+	pthread_mutex_destroy(&g_opt.mt);
+
+	/* free command buffer and list */
+	if (!g_opt.cmd_list)
+		return;
+
+	for (i = 0; i < g_opt.num_of_cmds; i++)
+		free(g_opt.cmd_list[i].cmd);
+
+	free(g_opt.cmd_list);
+}
+
+/*
+ * parse options
+ */
+
+static int parse_options(const char *timeout, const char *num_of_forks, const char *cmd_file)
+{
+	FILE *cmd_fp;
+
+	/* get timeout */
+	g_opt.timeout = (int)strtoul(timeout, NULL, 10);
+	if (g_opt.timeout < 0 || g_opt.timeout > 60) {
+		fprintf(stderr, "Invalid timeout value '%d'. Please specify a value in [0 ~ 60] seconds.\n",
+				g_opt.timeout);
+		return -1;
+	}
+
+	/* get command list from file */
+	cmd_fp = fopen(cmd_file, "r");
+	if (!cmd_fp) {
+		fprintf(stderr, "Could not open command file '%s' for reading\n", cmd_file);
+		exit(1);
+	}
+
+	while (!feof(cmd_fp)) {
+		char *line = NULL;
+		size_t len = 0, ret;
+
+		/* get command line */
+		ret = getline(&line, &len, cmd_fp);
+		if (ret < 0) {
+			if (line)
+				free(line);
+			break;
+		}
+		line[ret] = '\0';
+
+		/* chop endline */
+		if (strstr(line, "\r\n"))						/* windows format */
+			line[strlen(line) - 2] = '\0';
+		else if (line[strlen(line) - 1] == '\n')
+			line[strlen(line) - 1] = '\0';
+
+		/* check the length of line and comment */
+		if (strlen(line) == 0 || line[0] == '#') {
+			free(line);
+			continue;
+		}
+
+		/* add new command to list */
+		g_opt.cmd_list = (struct spawn_cmd *) realloc(g_opt.cmd_list,
+								(g_opt.num_of_cmds + 1) * sizeof(struct spawn_cmd));
+		if (!g_opt.cmd_list) {
+			fprintf(stderr, "Out of memory!\n");
+			free(line);
+			break;
+		}
+
+		memset(&g_opt.cmd_list[g_opt.num_of_cmds], 0, sizeof(struct spawn_cmd));
+		g_opt.cmd_list[g_opt.num_of_cmds++].cmd = line;
+	}
+
+	/* close file */
+	fclose(cmd_fp);
+
+	/* check count of commands */
+	if (g_opt.num_of_cmds == 0) {
+		fprintf(stderr, "Invalid command file '%s'\n", cmd_file);
+		exit(1);
+	}
+
+	/* get forks count */
+	g_opt.num_of_forks = strtoul(num_of_forks, NULL, 10);
+	if (g_opt.num_of_forks <= 0 || g_opt.num_of_forks > g_opt.num_of_cmds) {
+		fprintf(stderr, "The number of forks should be value in [1...%d]\n", g_opt.num_of_cmds);
+		exit(1);
```

Comment:
I don't think it's an error if number of forks is less than commands. Either silently reduce `num_of_forks` or allow the spawn loop to catch this case.